lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp

```diff
@@ -431,6 +434,93 @@ bool circt::firrtl::fromJSON(json::Value &value, StringRef circuitTarget,
   return true;
 }
 
+bool circt::firrtl::fromOMIRJSON(json::Value &value, StringRef circuitTarget,
+                                 llvm::StringMap<ArrayAttr> &annotationMap,
+                                 json::Path path, CircuitOp circuit) {
+  auto *context = circuit.getContext();
+
+  // The JSON value must be an array of objects.  Anything else is reported as
+  // invalid.
+  auto *array = value.getAsArray();
+  if (!array) {
+    path.report(
+        "Expected OMIR to be an array of nodes, but found something else.");
+    return false;
+  }
+
+  // Build a mutable map of Target to Annotation.
+  SmallVector<Attribute> omnodes;
+  for (size_t i = 0, e = (*array).size(); i != e; ++i) {
+    auto *object = (*array)[i].getAsObject();
+    auto p = path.index(i);
+    if (!object) {
+      p.report("Expected OMIR to be an array of objects, but found an array of "
+               "something else.");
+      return false;
+    }
+
+    // Manually built up OMNode.
+    NamedAttrList omnode;
+
+    // Validate that this looks like an OMNode.  This should have three fields:
+    //   - "info": String
+    //   - "id": String that starts with "OMID:"
+    //   - "fields": Array<Object>
+    // Fields is optional and can be either an array or not there.  If not
+    // there, then construct an empty array for it.
+    auto maybeInfo = object->getString("info");
+    if (!maybeInfo) {
+      p.report("OMNode missing mandatory member \"info\" with type \"string\"");
+      return false;
+    }
+    auto maybeID = object->getString("id");
+    if (!maybeID || !maybeID.getValue().startswith("OMID:")) {
+      p.report("OMNode missing mandatory member \"id\" with type \"string\" "
+               "that starts with \"OMID:\"");
+      return false;
+    }
+    auto *maybeFields = object->get("fields");
+    if (maybeFields && !maybeFields->getAsArray()) {
+      p.report("OMNode has \"fields\" member with incorrect type (expected "
+               "\"array\")");
+      return false;
+    }
+    Attribute fields;
+    if (!maybeFields)
+      fields = ArrayAttr::get(context, {});
+    else
+      fields = convertJSONToAttribute(context, *maybeFields, p.field("fields"));
+
+    omnode.append("info", convertJSONToAttribute(context, *object->get("info"),
+                                                 p.field("info")));
```

Comment:
The examples I've seen so far always have "info" set to a FIR location tag like `@[SRAM.scala 42:9001]`. If that's all it's ever going to be, we might want to convert this to a `LocationAttr` so we have an easier time manipulating OMIR in firtool and updating locations where necessary?

Reply:
Yes, the info is supposed to be a Chisel3 `SourceInfo`. We can do the conversion here. Great idea. (I was worried about how to verify if this string looked okay, but we can just parse it into a location. 💯 )

Reply:
I pulled all the logic related to `StringRef -> LocationAttr` into a utility function, `circt::firrtl::maybeStringToLocation`.  This function has a pretty verbose API that can be cleaned up later.  This will enable fused locations to be parsed out of the OMIR stuff.  I'm not aware of this happening, but this should future proof it. 😄